### PR TITLE
sampling module improvements

### PR DESCRIPTION
**Pull request by sraizada-tt**
This PR enhances the sampling module to support larger batch sizes beyond 32 and introduces multi-row data-parallel sampling (sampling_dp > 1) for models that shard users across mesh rows. The changes enable models like GPT-OSS on Galaxy (4×8 mesh) to sample 128 users (4 rows × 32 users per row) independently.

Changes:

Extended sampling to support batch sizes that are multiples of 32 (previously hardcoded to 32)
Added sampling_dp parameter to enable row-wise sampling parallelism across mesh rows
Refactored LogProbsCalculator from models/common/utils.py to models/common/sampling/tt_log_probs.py
Created models/common/sampling/_utils.py for shared helper functions (clamp, is_default_value, filter_none)
Added backward compatibility imports in models/common/utils.py
Updated SeedManager and format_sampling_params to support configurable batch sizes
Added force_argmax_sampling property to TTSampling
Updated documentation (README.md, new AGENTS.md)

llama 8b models ci: https://github.com/tenstorrent/tt-shield/actions/runs/22172504973
 llama 70b models ci: https://github.com/tenstorrent/tt-shield/actions/runs/22186164285
 Qwen3-32b models CI: https://github.com/tenstorrent/tt-shield/actions/runs/22302203153
 vllm nighlty: https://github.com/tenstorrent/tt-metal/actions/runs/22226398573
 
 Issue tracking unit tests bring up: https://github.com/tenstorrent/tt-metal/issues/38316

### CODE REVIEW

**Review by gwangTT — APPROVED** [review]

**Review by ssanjayTT — APPROVED** [review]

**Review by handrewsTT — APPROVED** [review]

File: models/common/sampling/AGENTS.md
Comment (mtairum) [inline models/common/sampling/AGENTS.md:1]: I odn't think this was supposed to be merged :)

**Review by mtairum — CHANGES_REQUESTED** [review]
Can you also provide a more detailed PR description?

**Review by mtairum — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -411,15 +354,16 @@ def prefill_forward_text(
 
             if sampling_enabled:
                 sampling_executed = True
+                sampling_dp = getattr(self.model[model_id], "sampling_dp", 1)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:357]: should there be a safeguard in case a model accidentally doesn't set that attr?

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -503,7 +447,7 @@ def prefill_forward_text(
                 idx = res["idx"]
                 last_token_idx = res["last_token_idx"]
                 model_id = res["model_id"]
-                num_cached_tokens = int(start_pos[elem_idx]) if start_pos is not None else 0
+                num_cached_tokens = int(start_pos[idx]) if start_pos is not None else 0
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:453]: any particular reason for this? seems ok but was just curious since elem_idx and idx should be same right?

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -703,43 +647,53 @@ def decode_forward(
...
-                chunked_fields[field] = self._chunk_sampling_param(val)
+            # sampling_dp may differ from data_parallel for models that internally
+            # shard users across mesh rows (users_row_sharded) — each row samples
+            # 32 users independently, so sampling params must be chunked by the
+            # number of rows even though data_parallel=1 for the forward pass.
+            sampling_dp_values = [getattr(self.model[i], "sampling_dp", 1) for i in range(self.data_parallel)]
+            assert (
+                len(set(sampling_dp_values)) == 1
+            ), f"All model instances must have the same sampling_dp, got {sampling_dp_values}"
+            sampling_dp = max(self.data_parallel, sampling_dp_values[0])
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:664]: may break when we have models that are both data_parallel > 1 and shard users across mesh rows right? don't need to support now but may warrant a comment

**Review by skhorasganiTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py:357]: The `getattr` default of 1 is intentionally the safeguard. Most models (e.g. Llama 8B) don't define `sampling_dp` because they don't row-shard users. Only models like Galaxy 70B / GPT-OSS that internally shard across mesh rows set this attribute. Default=1 means "standard, no row sharding."

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py:453]: Today they're always equal (every iteration of the first loop unconditionally appends one result), so this is a no-op change. But `idx` is semantically the right thing to use — it's the actual batch index stored in the result dict, while `elem_idx` is just the position in the results list. Using `idx` is more robust if the first loop ever gains a `continue` path.

File: models/tt_transformers/tt/generator.py (hunk 3 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py:664]: Good catch — added a comment clarifying this assumption in 4b2022a5fb:

```python
# NOTE: This assumes data_parallel and sampling_dp are mutually exclusive
# (one is always 1). If a future model needs both DP>1 and row-sharded
# sampling, this should become data_parallel * sampling_dp_values[0].
sampling_dp = max(self.data_parallel, sampling_dp_values[0])
```

File: models/common/sampling/tt_log_probs.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        self.common_args = filter_none(
+            {
+                "sub_core_grids": sub_core_grids,
+            }
+        )
+
+        # CCL introspection (same pattern as TTSampling)
+        self._line_all_gather = getattr(self.tt_ccl, "line_all_gather", None)
+        self._line_all_gather_supports_buffer_key = False
+        if callable(self._line_all_gather):
```
Comment (djordje-tt) [inline models/common/sampling/tt_log_probs.py:49]: What is this? Could we somehow else determine if line_all_gather supports buffer key?

Comment (sraizada-tt) [reply models/common/sampling/tt_log_probs.py:49]: all gather functions for different models

Comment (sraizada-tt) [reply models/common/sampling/tt_log_probs.py:49]: will have to make another attribute and pass it all the way through

**Comment by tchedaTT** [comment]
Please resolve copilot comments if they've been addressed or comment under them why they don't apply

File: models/common/sampling/tt_log_probs.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+        num_devices = self.mesh_device.get_num_devices()
+
+        # Determine the TP dimension: for 2D meshes, logits are sharded across
+        # the larger dimension (TP). For 1D or single-device, use all devices.
+        if self.cluster_shape[0] > 1 and self.cluster_shape[1] > 1:
+            # 2D mesh: TP axis is the larger dimension
+            tp_axis = 0 if self.cluster_shape[0] >= self.cluster_shape[1] else 1
+            num_devices_for_sharding = self.cluster_shape[tp_axis]
+            self._all_gather_cluster_axis = tp_axis
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/common/sampling/tt_log_probs.py:67]: The heuristic is correct for all current mesh configurations — vocab is always sharded across the larger dimension (Galaxy [8,4] uses axis 0 with 8 devices, [4,8] uses axis 0 with 4 devices for TP). That said, `TTSampling` does have a configurable `sampling_all_gather_axis` parameter. We could thread that through to `LogProbsCalculator` if a future configuration breaks the heuristic. For now, the comment on lines 61-62 documents the assumption.

File: models/common/sampling/tt_log_probs.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+        # Get logits for each user on each chip based on local index
+        selected_logits_tensor = ttnn.gather(logits_tensor, dim=3, index=remainder_tensor, **self.common_args)
+
+        # convert to ROW_MAJOR_LAYOUT due to memory clobbering which affects all ttnn.reshape ops with TILE_LAYOUT
+        selected_logits_tensor = ttnn.to_layout(selected_logits_tensor, ttnn.ROW_MAJOR_LAYOUT, **self.common_args)
+        batch_vol_s = selected_logits_tensor.shape[2] * selected_logits_tensor.shape[3]
+        selected_logits_tensor = ttnn.reshape(selected_logits_tensor, (1, 1, 1, batch_vol_s), **self.common_args)
+        selected_logits_tensor = ttnn.to_layout(selected_logits_tensor, ttnn.TILE_LAYOUT, **self.common_args)
+        # Compare mask to chip_ids tensor and select correct positions for each user on all chips inplace
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/common/sampling/tt_log_probs.py:240]: The batch_size here is always set to max_batch_size and tensors are always padded to that size before reaching this code — there's no runtime batch size variation. The mask and incoming tensors will always match. Adding a shape check would be dead code.

File: models/common/sampling/tt_sampling.py [hunk 7]
```diff
@@ -101,6 +100,12 @@ def __init__(
         self.sub_core_grid_topk = getattr(args, "sub_core_grid_topk", None)
         self.start_core = getattr(args, "start_core", ttnn.CoreCoord(0, 0))
 
+        # sampling_dp > 1 when multiple mesh rows each sample 32 users independently
+        # (e.g. GPT-OSS on [4,8] Galaxy: 4 rows × 32 users = 128 total users)
+        self._sampling_dp = getattr(args, "sampling_dp", 1)
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/common/sampling/tt_sampling.py:123]: sampling_dp is set by model code we control, not external user input. The getattr default of 1 already handles the missing case. Adding runtime type/value validation for internal-only attributes is unnecessary defensive coding.

File: models/common/sampling/tt_sampling.py [hunk 8]
```diff
@@ -45,21 +46,18 @@ class TTSampling(LightweightModule):
...
-        return all(value == default for value in values)
-
     def _is_force_argmax_sampling(self, k, p, temp):
         return (
             self._allow_force_argmax_sampling
-            and self._is_default_val(k, 1)
-            and self._is_default_val(p, 1.0)
-            and self._is_default_val(temp, 1.0)
+            and is_default_value(k, 1)
+            and is_default_value(p, 1.0)
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/common/sampling/tt_sampling.py:69]: This is not an inconsistency. `_is_force_argmax_sampling` checks the **caller's parameters** — `p=1.0` from the caller means "no top-p filtering," which combined with k=1 and temp=1.0 means pure argmax. The internal default of `p=0` at line 147 is a separate concern — it's the device kernel's representation for "top-p disabled." These are two different layers: caller semantics (p=1.0 = keep all tokens) vs device defaults (p=0 = no filtering). The README documents this distinction in the Pitfalls section.

File: models/common/utils.py [hunk 9]
```diff
@@ -5,7 +5,11 @@
 import torch.nn.functional as F
 from torch import Tensor
 
-import ttnn
+# Backward compatibility: filter_none moved to models.common.sampling._utils
+from models.common.sampling._utils import filter_none  # noqa: F401
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/common/utils.py:9]: These are intentional backward-compatibility re-exports — other code imports `filter_none` and `LogProbsCalculator` from `models.common.utils`. The `# noqa: F401` comments explicitly tell linters to allow the unused-looking imports.

File: models/common/utils.py [hunk 10]
```diff
@@ -5,7 +5,11 @@
...
+
+# Backward compatibility: LogProbsCalculator moved to models.common.sampling.tt_log_probs
+from models.common.sampling.tt_log_probs import LogProbsCalculator  # noqa: F401
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/common/utils.py:12]: Same as above — intentional backward-compatibility re-export with `# noqa: F401`.

File: models/tt_transformers/tt/generator.py [hunk 11]
```diff
@@ -418,14 +414,16 @@ def prefill_forward_text(
...
-                    _broadcast_formatted_sampling_params(sampling_params, idx), 32
+                    _broadcast_formatted_sampling_params(sampling_params, idx, slot_len=total_batch), total_batch
                 )
                 assert per_request_params is not None, "Sampling was executed but missing per-request sampling params"
                 _apply_prefill_sampling_state(
                     self.model[model_id],
                     sampling_params=per_request_params,
-                    prompt_tokens=prefill_ids[:, :seq_len].repeat(32, 1),
+                    prompt_tokens=prefill_ids[:, :seq_len].repeat(total_batch, 1),
                     empty_slots=[user_id % 32],
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply models/tt_transformers/tt/generator.py]: Could you add a comment why this why doesn't apply? It's not immediately obvious

[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: Added a comment in 5556959671:

```python
# empty_slots uses max_batch_size_per_model (not total_batch) because
# the seed manager operates on per-row slots (0..31).  When sampling_dp > 1
# the params are already broadcast across all rows by broadcast_sampling_params.
```

During prefill we handle one user at a time. `user_id % max_batch_size_per_model` gives the user's slot within the per-row batch (0-31). `broadcast_sampling_params` already replicates params across all `total_batch` slots, so the seed manager only needs to know which of the 32 per-row slots this user occupies.

**Comment by tchedaTT** [comment]
Could you document what is the purpose of force_argmax_sampling ?

**Comment by tchedaTT** [comment]
+ given we have non-trivial logic for when traces are valid, should we fully deprecate non-split tracing of sampling in ttt/ llama70b? AFAIK there's currently an always-on switch there, but it seems like there's no way to correctly trace sampling jointly with the model (the model doesn't know when the trace is still valid)

**Comment by sraizada-tt** [comment]
> When do we want to switch force_argmax_sampling on a per-batch level? It seems like with having to invalidate the trace this would be slower than just running regular sampling, not to mention potentially running into the bugs with trace capture that require us to pre-warm all traces

I think it is set at a model level, it was added on the stable branch PR because llama 8b needed it to meet beanchmarks

**Review by tchedaTT — APPROVED** [review]
Approving pending the CI run requested by @djordje-tt
